### PR TITLE
Data source property types for PW [for 8.7]

### PR DESCRIPTION
**Pull request by r0b1n**

### CODE REVIEW

File: content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md [hunk 1]
```diff
@@ -326,6 +332,20 @@ then the Studio Pro UI for the component appears like this:
...
+
+```xml
+<property key="content" type="widgets" required="false" dataSource="myDataSource">
+	<caption>Content</caption>
+	<category>General</category>
+	<description>Widgets using data source</description>
+</property>
+```
+
+Where the `myDataSource` is a [`datasource`](#datasource) property, the Studio Pro UI for the component appears like this:
```
Comment (ConnorLand) [inline content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: @r0b1n I don't fully understand how this line works, especially in conjunction with the "when the component is defined" line above it. Usually line 1 is "when the component is as follows" and line 2 is just "then it appears in the UI this way". Can you help clarify please?

File: content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md [hunk 2]
```diff
@@ -371,7 +391,7 @@ Then the Studio Pro UI for the property appears like this:
 
 ### 4.2 TextTemplate{#texttemplate}
 
-`TextTemplate` allows a user to configure translatable text template similar to the [Caption](/refguide/text#caption) of a text widget. The interpolated string will be passed to the client component as `DynamicValue<string>`.
+The TextTemplate property allows a user to configure translatable text template similar to the [Caption](/refguide/text#caption) of a text widget. The interpolated string will be passed to the client component as `DynamicValue<string>`.
```
Comment (ConnorLand) [inline content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: Should all properties be uppercase, or only ones made of two words? How do they appear in the code @r0b1n ?

Also, currently, all properties are not in code tics. Their types do get code tics in the XML attributes section, however, and these types are letter for letter identical to the properties. Is this a problem?

File: content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md (hunk 1 above)
Comment (r0b1n) [reply content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: Rephrased

File: content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md (hunk 2 above)
Comment (r0b1n) [reply content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: I think in this case it is fine to use *text template*, just as a normal text. Using `TextTemplate` is not technically correct, it is `textTemplate`. Whenever a text is in code tics it shouldn't change, capitalization rules shouldn't apply to it. Maybe we should go with *Text template* for the header as well.

I would use properties in tics everywhere, but it is sub-optimal from the stylistic point of view. As we are using non-ticked versions to name other properties like *expression*, *decimal*, etc., we can go with normal two words (*text template*) for this one as well.

Comment (ConnorLand) [reply content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: Ok. I am going to use this PR as an opportunity to go through all my API docs. I'm going to reexamine the way certain terms use code tics, and I'm going to reevaluate combining code tics and hyperlinks.

**Comment by jordanch** [comment]
@r0b1n Does "datasource" need to be added as a type to 1.1.2 Type list?

File: content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md [hunk 3]
```diff
@@ -175,6 +175,46 @@ export type ImageValue = WebImage | NativeImage;
...
+    setAmount: (Option<number>) => void;
+    offset: number;
+    setOffset: (number) => void;
+}
+```
+
+
+When a `datasource` property with `isList="true"` is configured for a widget, the client component gets a list of objects represented as a `ListValue`. This type allows detailed access to a data source, and enables control over the amount and offset of items represented in the list.
+
+However it is not possible to access domain data directly from `ListValue`, as every object is represented only by GUID in the `items` array. Instead, a list of items may be used in combination with other properties so long as a `datasource` property is tied to with the `dataSource` attribute on that property.
```
[the remark this answers is not in the document]
Comment (r0b1n) [reply content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md]: Rephrased

File: content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md [hunk 4]
```diff
@@ -175,6 +175,46 @@ export type ImageValue = WebImage | NativeImage;
...
+    totalCount: Option<number>;
+    amount: number;
+    setAmount: (Option<number>) => void;
+    offset: number;
+    setOffset: (number) => void;
+}
+```
+
+
+When a `datasource` property with `isList="true"` is configured for a widget, the client component gets a list of objects represented as a `ListValue`. This type allows detailed access to a data source, and enables control over the amount and offset of items represented in the list.
```
[the remark this answers is not in the document]
Comment (r0b1n) [reply content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md]: ✅

File: content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md [hunk 5]
```diff
@@ -175,6 +175,46 @@ export type ImageValue = WebImage | NativeImage;
...
+export interface ListItem {
+    id: GUID;
+}
+
+export interface ListValue {
+    status: ValueStatus;
+    items: ListItem[];
+    totalCount: Option<number>;
+    amount: number;
+    setAmount: (Option<number>) => void;
```
[the remark this answers is not in the document]
Comment (r0b1n) [reply content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md]: ✅

File: content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md [hunk 6]
```diff
@@ -175,6 +175,46 @@ export type ImageValue = WebImage | NativeImage;
...
+```ts
+export interface ListItem {
+    id: GUID;
+}
+
+export interface ListValue {
+    status: ValueStatus;
+    items: ListItem[];
+    totalCount: Option<number>;
+    amount: number;
```
[the remark this answers is not in the document]
Comment (r0b1n) [reply content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md]: ✅

File: content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md [hunk 7]
```diff
@@ -185,6 +185,48 @@ export interface FileValue {
...
+When a `datasource` property with `isList="true"` is configured for a widget, the client component gets a list of objects represented as a `ListValue`. This type allows detailed access to a data source, and enables control over the limit and offset of items represented in the list.
+
+However it is not possible to access domain data directly from `ListValue`, as every object is represented only by GUID in the `items` array. Instead, a list of items may be used in combination with other properties, for example with a property of type [`widgets`](property-types-pluggable-widgets#widgets). The `dataSource` attribute on that property should refference a `datasource` property.
+
+For clarity, consider the following example using `ListValue` together with the `widgets` property type. When the `widgets` property named `myWidgets` is configured to be tied to a `datasource` named `myDataSource`, the client component props appear as follows:
+
+```ts
+interface MyListWidgetsProps {
+    myDataSource: ListValue;
+    myWidgets: (ListItem) => ReactNode;
```
Comment (mijay) [inline content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md]: `(i: ListItem) => ReactNode`

File: content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md [hunk 8]
```diff
@@ -561,6 +581,43 @@ Then the Studio Pro UI for the property appears like this:
 
 ![](attachments/widget-property-types/file.png)
 
+
+### 4.7 Datasource {#datasource}
```
Comment (mijay) [inline content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: There is also a PW-vs-CW table, which we should now change to say "yes, ds are supported in PW".

File: content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md (hunk 7 above)
Comment (r0b1n) [reply content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md]: ✔️

File: content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md (hunk 8 above)
Comment (r0b1n) [reply content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: ✔️

File: content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md (hunk 7 above)
Comment (mijay) [reply content/apidocs-mxsdk/apidocs/client-apis-for-pluggable-widgets.md]: Few similar one in `ListValue` itself.

**Comment by r0b1n** [comment]
Wont make it till 8.6, postponed to 8.7

File: content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md [hunk 9]
```diff
@@ -561,6 +581,43 @@ Then the Studio Pro UI for the property appears like this:
...
+
+The datasource property allows widgets to work with object lists. The client component will receive value prop of type [`ListValue`](client-apis-for-pluggable-widgets#listvalue) and may be used with the [`widgets`](#widgets) property.
+
+{{% alert type="info" %}}
+Support for the datasource property type was introduced in Mendix 8.6.
```
Comment (mijay) [inline content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: 8.7

File: content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md [hunk 10]
```diff
@@ -295,26 +296,32 @@ Then the Studio Pro UI for the component appears like this:
 
 ### 3.3 Widgets {#widgets}
 
-The widgets property allows a user to place multiple widgets inside a pluggable widget, similar to the content of a [container](/refguide/container) widget. It is passed as a `ReactNode` prop to a client component.
+The widgets property allows a user to place multiple widgets inside a pluggable widget, similar to the content of a [container](/refguide/container) widget. It is passed as a `ReactNode` prop to a client component if a `dataSource` attribute is not specified or if attribute is specified, but the data source is not configured by the user. Otherwise it is passed as a function that expects a `ListItem` and returns a `ReactNode`: `(item: ListItem) => ReactNode`. For more information, see the [datasource](#datasource) section below.
```
Comment (r0b1n) [inline content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: ListItem => ObjectItem

**Comment by mijay** [comment]
Shall we un-wip this @r0b1n?

**Comment by Adam-Dupaski** [comment]
@ConnorLand Please resolve these conflicts, finish the WIP, and remove the "WIP" label so this can be merged for the 8.7.0 release tomorrow.

**Comment by ConnorLand** [comment]
I will do this today. GitHub in the browser told me the conflicts are too complex to be fixed in the browser, so I’ve reached out to Roman asking him to fix them on his end.

File: content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md (hunk 9 above)
Comment (r0b1n) [reply content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: ✔️

File: content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md [hunk 11]
```diff
@@ -295,26 +296,32 @@ Then the Studio Pro UI for the component appears like this:
...
 
 {{% alert type="info" %}}
 This property type was introduced in Mendix 8.3.
 {{% /alert %}}
 
+{{% alert type="info" %}}
+Support for the `dataSource` attribute was introduced in Mendix 8.6.
```
[the remark this answers is not in the document]
Comment (r0b1n) [reply content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: ✅

File: content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md [hunk 12]
```diff
@@ -561,6 +581,43 @@ Then the Studio Pro UI for the property appears like this:
 
 ![](attachments/widget-property-types/file.png)
 
+
+### 4.7 Datasource {#datasource}
+
+The datasource property allows widgets to work with object lists. The client component will receive value prop of type [`ListValue`](client-apis-for-pluggable-widgets#listvalue) and may be used with the [`widgets`](#widgets) property.
```
[the remark this answers is not in the document]
Comment (r0b1n) [reply content/apidocs-mxsdk/apidocs/property-types-pluggable-widgets.md]: ✅

**Pull request merged**